devito/passes/iet/dtypes.py

```diff
@@ -0,0 +1,58 @@
+import numpy as np
+import ctypes
+
+from devito.ir import FindSymbols, Uxreplace
+
+__all__ = ['lower_complex']
+
+
+def lower_complex(iet, lang, compiler):
+    """
+    Add headers for complex arithmetic
+    """
+    # Check if there is complex numbers that always take dtype precedence
+    types = {f.dtype for f in FindSymbols().visit(iet)
+             if not issubclass(f.dtype, ctypes._Pointer)}
+
+    if not any(np.issubdtype(d, np.complexfloating) for d in types):
+        return iet, {}
+
+    lib = (lang['header-complex'],)
+
+    metadata = {}
+    if lang.get('complex-namespace') is not None:
+        metadata['namespaces'] = lang['complex-namespace']
+
+    # Some languges such as c++11 need some extra arithmetic definitions
+    if lang.get('def-complex'):
+        dest = compiler.get_jit_dir()
+        hfile = dest.joinpath('complex_arith.h')
+        with open(str(hfile), 'w') as ff:
+            ff.write(str(lang['def-complex']))
+        lib += (str(hfile),)
+
+    iet = _complex_dtypes(iet, lang)
+    metadata['includes'] = lib
+
+    return iet, metadata
+
+
+def _complex_dtypes(iet, lang):
+    """
+    Lower dtypes to language specific types
+    """
+    mapper = {}
+
+    for s in FindSymbols('indexeds').visit(iet):
```

Comment:
You can chain the two FindSymbols by

```
for i in FindSymbols('indexeds|symbolics').visit(iet):
     if i.dtype in lang['types']:
           ....
```

A comprehension might be usable too, but it might rather clutter the code, up to you

Reply:
Yes, leftovers had different rebuilds because of indexes, will merge